keps/sig-auth/3130-kms-observability/README.md

```diff
@@ -0,0 +1,224 @@
+
+### Monitoring Requirements
+
+###### How can someone using this feature know that it is working for their instance?
+
+- [x] Other (treat as last resort)
+  - Details: Logs in kube-apiserver, kms-plugin and KMS will be logged with the corresponding UID.
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+There should be no impact on the SLO with this change.
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+- [x] Other (treat as last resort)
+  - Details: Logs in kube-apiserver, kms-plugin and KMS will be logged with the corresponding UID.
+
+### Dependencies
+
+###### Does this feature depend on any specific services running in the cluster?
+
+No.
+
+### Scalability
+
+###### Will enabling / using this feature result in any new API calls?
+
+No.
+
+###### Will enabling / using this feature result in introducing new API types?
+
+No.
+
+###### Will enabling / using this feature result in any new calls to the cloud provider?
+
+No.
+
+###### Will enabling / using this feature result in increasing size or count of the existing API objects?
+
+No.
+
+###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?
+
+No.
+
+###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?
+
+No.
+
+### Troubleshooting
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+- Secret data encryption with external kms-plugin is unavailable
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+We considered using the AuditID from the kube-apiserver request that generated the envelope operation. This approach has the following drawbacks:
+
+1. AuditID can be configured by the user with the `Audit-ID` header in the API server request. Multiple requests can be sent to the kube-apiserver with the same Audit-ID.
+2. Not all API server requests will generate an envelope operation. The API server caches DEKs and for the DEK that's available in the cache, the kube-apiserver will not generate an envelope operation.
```

Comment:
I don't think that is an issue. When troubleshooting I don't think we will care about the operations that are not hitting the kms since they shouldn't fail in the first place and they won't have any impact on the kms.

That said, I don't think hitting the cache is worth logging, but if we think it is useful to keep track of how much the cache is hit, then a metric would be more appropriate.

Reply:
> I don't think that is an issue. When troubleshooting I don't think we will care about the operations that are not hitting the kms since they shouldn't fail in the first place and they won't have any impact on the kms.

With using auditID and having audit logging, this is a drawback because there is no corresponding entry in kms for the UID. This is under the alternatives that we considered but not the actual proposal. Hope that clarifies it?

Reply:
To me, that is completely normal since the kms provider will not be part of this communication. The information that there is no corresponding entry in kms for the UID can be understood as kms can be ruled out from the potential cause of the problem and that the cache needs to be further investigated.